src/ast2ram/seminaive/ConstraintTranslator.cpp

```diff
@@ -29,10 +28,9 @@
 #include "ram/ProvenanceExistenceCheck.h"
 #include "ram/UndefValue.h"
 
-namespace souffle::ast2ram {
+namespace souffle::ast2ram::seminaive {
 
-Own<ram::Condition> ConstraintTranslator::translate(const TranslatorContext& context,
-        SymbolTable& symbolTable, const ValueIndex& index, const ast::Literal* lit) {
+Own<ram::Condition> ConstraintTranslator::translateConstraint(const ast::Literal* lit) {
```

Comment:
```suggestion
Own<ram::Condition> ConstraintTranslator::translateConstraint(const ast::Literal& lit) {
```

This would enforce the invariant at the type level, and the assert would be unnecessary.

Reply:
This is true locally, but the literal is originally a pointer, so the assertion has to be there at some point when it is first dereferenced. The assertion can be moved to the call-site, but the non-nullptr requirement is needed whenever this (and the value translator) are called anyway. In this case, passing in a pointer and checking it in one location was simpler/easier than checking it at each callpoint.

Reply:
Makes sense.  However, as far as I can see the vast majority of these will come from things like `atom->getArguments()`, and they are produced by `toPtrVector`.

Not that this should be done now, but how about replacing the `std::vector<T*>` return types with `std::vector<std::reference_wrapper<T>>`?